### ISSUE TITLE
Inefficient use of pmap in jax numpyro sampler?

### ISSUE THREAD

**Issue by twiecki** [pymc-devs/pymc#4288]
```
/Users/twiecki/miniconda3/envs/pymc3theano/lib/python3.8/site-packages/jax/interpreters/xla.py:624: UserWarning: The jitted function _sample includes a pmap. Using jit-of-pmap can lead to inefficient data movement, as the outer jit does not preserve sharded data representations and instead collects input and output arrays onto a single device. Consider removing the outer jit unless you know what you're doing. See https://github.com/google/jax/issues/2926.
  warn(f"The jitted function {fun.__name__} includes a pmap. Using "
```

**Comment by junpenglao** [comment]
The pmap and jit happens inside of Numpyro so I am not sure there is anything we can do about it.

**Comment by twiecki** [comment]
from Du Phan: "It seems to me that jit + pmap is not needed. You can run MCMC(NUTS(potential_fn)) with a batch (num_chains) of init_params."

**Comment by neerajprad** [comment]
Also, to your earlier point, I will be very surprised if `jit(pmap..)` is faster since pmap will run the computation on different devices, so I don't see much scope for graph based optimization. Running multiple chains will use `pmap(jit(..))` under the hood, and you can also try using `chain_method='vectorized'` (which uses `lax.map`) to compare performance (I think compilation might be faster on one vs the other).

**Comment by roblem** [comment]
Using `pm.sampling_jax.sample_tfp_nuts`, with multiple gpu's, causes errors if chains > number of devices since as pointed out above, jax will distribute each batch to a separate device.  I changed the jax wrapper [here](https://github.com/pymc-devs/pymc3/blob/aedc8e9006c7e672e461774a5ee095562405c99c/pymc3/sampling_jax.py#L56) to `@jax.vmap` and it runs (batched) on a single gpu which is what you'd want in most cases (`@jax.jit` throws an error).  

I also tried numpyro using  `chain_method='vectorized'`  [here](https://github.com/pymc-devs/pymc3/blob/aedc8e9006c7e672e461774a5ee095562405c99c/pymc3/sampling_jax.py#L159) but it hangs after sampling (the progress bar is at 100%).  With the `vectorized` setting, it does seem to be doing batch sampling on a single gpu (and faster than the tfp sampling method).

**Comment by twiecki** [comment]
@roblem Thanks for posting. Any other experiences you can share in regards to speed, model size, data size or other hickups?

**Comment by roblem** [comment]
I've mostly been looking into modifications to `pm.sampling_jax.sample_tfp_nuts` in order to do batching on a single gpu in an analogous way to tensorflow_probability.  To do that, I've created a fork that wraps `_sample` in both `jax.jit` and `jax.pmap` and this seems to fix the issue described above that the number of chains has to be less than or equal to the number of devices due to the use of `jax.pmap`.  I've created a test version of `pm.sampling_jax` [here](https://github.com/roblem/pymc3/blob/master/pymc3/sampling_jax.py) that allows for side-by-side comparisons between the current jax sampler and one that batches better as described above (it is called `pm.sampling_jax.sample_tfp_nuts_jit_vmap`).  

I love where the project is going with this, and have some observations about the new jax wrapper on gpus:
1. I believe all sampling is occurring as float64 which is going to be the slowest way to sample on gpus.  I tried casting things as float32, but steps sizes in the nuts sampler persisted with float64 and sampling would error out.
2. Sampling times on gpu are disappointing.  On my workstation, I have amd hardware and have a working jax/tensorflow install for that (on AMD ROCm) and what I am seeing in terms of run times is consistent with what I've seen in pure tensorflow/tensorflow probability: Nuts is slow on GPU (but can be especially slow on AMD hardware).  I've also ran the code on Nvidia hardware (on a google colab worksheet linked below) and see similar results relative to CPU.  CPU (even without xla compile from Jax) for Nuts sampling is going to be faster for lots of models.  As an FYI, other samplers on tensorflow_probability like basic metropolis-hastings is REALLY fast on GPU.  So it must be the nature of the step size and leapfrogging calculations on gpu....  maybe numpyro is better?
4. A very casual "eye-balling" of my gpu activity while sampling shows long drops in activity (5-10 seconds) between tuning and inference steps and at the end of sampling waiting for the chain to be returned.  I think this adds at least 15-20 seconds of overhead whether on Nvidia or AMD.  Note, this is additional to the initial compile.  I wonder if this is lots of IO activity moving gpu memory contents around.
5. Note: I haven't tested `pm.sampling_jax.sample_tfp_nuts_jit_vmap` on systems only having cpu's (and how jax xla compile would work).
3. Just judging by the progress bar on numpyro, it seemed to be sampling (and tuning) quite fast, but it would really hang after sampling for long periods before returning results.  There are also bugs/limitations where [it hangs if chains = 1](https://github.com/pymc-devs/pymc3/issues/4383), and using `pm.sampling_jax.sample_tfp_numpyro` with chains > number of hardware devices leads to sequential chain calculations which is inefficient.  I created a `pm.sampling_jax.sample_numpyro_nuts_vmap` that adds a function argument for sampling to enable the `chain_method=vectorized` option and sampling works, but it is either extremely slow or it hangs. 

Here are some model run times (note: all nvidia times were from [this shared worksheet on google colab](https://colab.research.google.com/drive/1mKIZC6rSnDhvs1t1sAzs-X4zJjDF6PFs?usp=sharing)).  Model 1 is a simple regression model with 1000 observations and 50 parameters, and Model 2 is the Radon Model from [here](https://docs.pymc.io/notebooks/GLM-hierarchical-jax.html) (thanks @twiecki).  When times are listed as "Error" below it is due to `pm.sampling_jax.sample_tfp_nuts` not allowing for chains > hardware devices.

|Model |Chains|	Hardware|Function Call|	Wall Time|
| ----- | ----| -------|----------|-------|
|Model 1: Toy OLS |	1	|CPU (20 cores)	|pm.sample|	39s|
|| 	1	|Radeon VII (AMD)	|pm.sampling_jax.sample_tfp_nuts	   |1m28s|
|| 	1	|Tesla T4 (Nvidia)	|pm.sampling_jax.sample_tfp_nuts	   |1m0s|
|| 	1	|Radeon VII (AMD)	|pm.sampling_jax.sample_tfp_nuts_jit_vmap  | 	1m43s|
|| 	1	|Tesla T4 (Nvidia)	|pm.sampling_jax.sample_tfp_nuts_jit_vmap  | 	1m7s|
|| 	50	|CPU (20 cores)	        |pm.sample	                           |33m35s|
|| 	50	|Radeon VII (AMD)	|pm.sampling_jax.sample_tfp_nuts	   |Error|
|| 	50	|Tesla T4 (Nvidia)	|pm.sampling_jax.sample_tfp_nuts	   |Error|
|| 	50	|Radeon VII (AMD)	|pm.sampling_jax.sample_tfp_nuts_jit_vmap  |3m53s|
|| 	50	|Tesla T4 (Nvidia)	|pm.sampling_jax.sample_tfp_nuts_jit_vmap  |3m38s|
|Model 2: Radon|	1|	CPU (20 cores)	|pm.sample	                           |14s|
|| 	1       |Radeon VII (AMD)	|pm.sampling_jax.sample_tfp_nuts	   |2m54s|
|| 	1	|Tesla T4 (Nvidia)	|pm.sampling_jax.sample_tfp_nuts	   |1m22s|
|| 	1	|Radeon VII (AMD)	|pm.sampling_jax.sample_tfp_nuts_jit_vmap  | 	2m49s|
|| 	1	|Tesla T4 (Nvidia)	|pm.sampling_jax.sample_tfp_nuts_jit_vmap  | 	1m53s|
|| 	50	|CPU (20 cores)	        |pm.sample	                           |             4m7s|
|| 	50	|Radeon VII (AMD)	|pm.sampling_jax.sample_tfp_nuts	   |Error|
|| 	50	|Tesla T4 (Nvidia)	|pm.sampling_jax.sample_tfp_nuts	   |Error|
|| 	50	|Radeon VII (AMD)	|pm.sampling_jax.sample_tfp_nuts_jit_vmap  | 	11m39s|
|| 	50	|Tesla T4 (Nvidia)	|pm.sampling_jax.sample_tfp_nuts_jit_vmap  | 	6m41s|

**Comment by PaoloRanzi81** [comment]
Hi @roblem  @twiecki, 

Outline: 
I am facing a similar issue by `sample_numpyro_nuts` function in `sampling_jax.py` by using 1 single GPU. Of course, I have manually modified `sampling_jax.py` by changing these two parameters: `progress_bar `+ `chain_method` . I am experiencing huge RAM memory errors by using PyMC3-Numpyro-JAX.

Background: 
I have a very complex hierarchical model (e.g. lots of hyper-priors etc.) in production. I have used the official release PyMC 3.10 + 32 vCPU + total of 450000 MCMC samples. The data-set is quite small (0.5 MB). It samples successfully within 3 Hrs: no error whatsoever. This is my benchmark against PyMC3-Numpyro-JAX.

I have tried the PyMC3-Numpyro-JAX version from PyMC3’s master branch together with 1 GPU. I have used the very same model + the very same data-set. Unfortunately, I have got repeatedly a nasty Out-Of-Memory (OOM) error. Particularly shocking, the fact that with `chain_method=“parallel”` + `chains=2` + `tune=5375` + `draws=1000`,  39 GB of RAM blew up with only a total of 10750 MCMC samples!  

I have already filed an issue in Numpyro ( https://github.com/pyro-ppl/numpyro/issues/899 ). By using a custom toy example, it looks like the problem is PyMC3 and not Numpyro. Two are the parameters which could be deemed guilty: 
-  `progress_bar = True` : Numpyro’s team (@fehiepsi) discovered that it is necessary to set it to “True” in order to avoid a monotonic increase of RAM memory use when using 1 GPU; 
- `chain_method = “parallel”` or `chain_method = “sequential”` : this are necessary for running the chain sequentially (since only 1 device is available). Theoretically, Numpyro should automatically revert to “sequential” whenever it finds only 1 available device. 

Major issue: 
Holding `progress_bar = True` , PyMC3 had several issues (even with just 10750 MCMC samples total): 
- `chain_method = “parallel”` : PyMC3 shows a monotonic increase of RAM (definitely an OOM error); 
- `chain_method = “sequential”` : PyMC3 does not even start; 
- `chain_method = “vectorized”` :  PyMC3 forces the use of CPUs overriding the GPU;

I have attached a table below about those testings. The red color are the bits that worry me. 
![pymc3_numpyro_OOM](https://user-images.githubusercontent.com/40964560/106323087-210c0c00-6277-11eb-82bb-0696643fb382.png)


Minor issue: 
I have also noticed when using CPUs by PyMC3-Numpyro-JAX, the vCPUs are not working full-throttle as I was used to see with libraries which used “joblib” as back-end (e.g. PyMC3 itself or Scikit-learn). Instead, with Numpyro-JAX the vCPUs work at a very low intensity. I have been used to spot the same phenomenon with TensorFlow when using vCPUs instead of accelerators. This should be not surprising since JAX should be very similar to TensorFlow. 
I am not even sure this is an issue. Indeed, it could be that the low intensity of JAX is as efficient as the higher intensity of “joblib” (e.g. same speed as “joblib”, but more electricity saved by JAX, thus is JAX overall superior?!?). 

To do: 
I have not tested the model neither by > 1 GPU nor 1 TPU. I would like to do it, as soon as I can put my hands on those accelerators. 

Thanks in advance for your help.

**Comment by junpenglao** [comment]
Thanks for reporting back your experience with the jax samplers! cc @ColCarroll and @brianwa84

**Comment by roblem** [comment]
I have multiple gpu's and haven't tested the memory leak problem, but noticed that on `numpyro` sampling in pymc3
1. chain_method = “parallel” sequentially executes on a single gpu when chains > # gpus leaving the other gpu idle
2. when chains == #gpus I get this error:  `RuntimeError: Unimplemented: Requested AllReduce not implemented on GPU; replica_count: 2; operand_count: 1; IsCrossReplicaAllReduce: 1; NCCL support: 0; first operand array element-type: S64`
3.  setting chain_method = “vectorized” does indeed force execution on cpu's

On tensorflow, using `pm.sampling_jax.sample_tfp_nuts_jit_vmap` will increase the intensity- whether measures as percent gpu usage or power draw- on a single gpu when batches (number of chains) is large enough

**Comment by twiecki** [comment]
These results and analyses here are very exciting, exactly the type of work that needs to be done. 

For reference also posting a NB by @junpenglao that compares the pymc3 and numpyro jaxpr: https://gist.github.com/junpenglao/c8b884797f950d1ef033ca69b253a4a0 and does some timings.

**Comment by brianwa84** [comment]
@roblem I would be curious to see the stack traces for the errors you were seeing on 50-batches with jax tfp nuts.

**Comment by roblem** [comment]
@brianwa84 the error happens for any case where chains > number of devices.  Using the toy ols model with 2 GPU devices:
```python
%%time
# Nuts tfp
with ols_model:
    trace_jax_tf = pm.sampling_jax.sample_tfp_nuts(10000, tune=1000, target_accept=0.8, chains=5)
``` 
gives this error (note  this is on jax/tensorflow under AMD ROCm but I know it errors out on Nvidia hardware with very similar errors (from [the colab notebook](https://colab.research.google.com/drive/1mKIZC6rSnDhvs1t1sAzs-X4zJjDF6PFs?usp=sharing) that should be accessible):
```
---------------------------------------------------------------------------
FilteredStackTrace                        Traceback (most recent call last)
<timed exec> in <module>

~/anaconda3/envs/pymc3-jax-roc/lib/python3.7/site-packages/pymc3/sampling_jax.py in sample_tfp_nuts(draws, tune, chains, target_accept, random_seed, model, num_tuning_epoch, num_compute_step_size)
    100     map_seed = jax.random.split(seed, chains)
--> 101     mcmc_samples, leapfrog_num = _sample(init_state_batched, map_seed)
    102 

FilteredStackTrace: ValueError: compiling computation that requires 5 logical devices, but only 2 XLA devices are available (num_replicas=5, num_partitions=1)

The stack trace above excludes JAX-internal frames.
The following is the original exception that occurred, unmodified.

--------------------

The above exception was the direct cause of the following exception:

ValueError                                Traceback (most recent call last)
<timed exec> in <module>

~/anaconda3/envs/pymc3-jax-roc/lib/python3.7/site-packages/pymc3/sampling_jax.py in sample_tfp_nuts(draws, tune, chains, target_accept, random_seed, model, num_tuning_epoch, num_compute_step_size)
     99     tic2 = pd.Timestamp.now()
    100     map_seed = jax.random.split(seed, chains)
--> 101     mcmc_samples, leapfrog_num = _sample(init_state_batched, map_seed)
    102 
    103     # map_seed = jax.random.split(seed, chains)

~/Downloads/jax/jax/_src/traceback_util.py in reraise_with_filtered_traceback(*args, **kwargs)
    137   def reraise_with_filtered_traceback(*args, **kwargs):
    138     try:
--> 139       return fun(*args, **kwargs)
    140     except Exception as e:
    141       if not is_under_reraiser(e):

~/Downloads/jax/jax/api.py in f_pmapped(*args, **kwargs)
   1572         out_axes_thunk=out_axes_thunk,
   1573         name=flat_fun.__name__, donated_invars=tuple(donated_invars),
-> 1574         global_arg_shapes=tuple(global_arg_shapes_flat))
   1575     return tree_unflatten(out_tree(), out)
   1576 

~/Downloads/jax/jax/core.py in bind(self, fun, *args, **params)
   1270   def bind(self, fun, *args, **params):
   1271     assert len(params['in_axes']) == len(args)
-> 1272     return call_bind(self, fun, *args, **params)
   1273 
   1274   def process(self, trace, fun, tracers, params):

~/Downloads/jax/jax/core.py in call_bind(primitive, fun, *args, **params)
   1724     if top_trace is None:
   1725       with new_sublevel():
-> 1726         outs = primitive.impl(fun, *args, **params)
   1727     else:
   1728       tracers = map(top_trace.full_raise, args)

~/Downloads/jax/jax/interpreters/pxla.py in xla_pmap_impl(fun, backend, axis_name, axis_size, global_axis_size, devices, name, in_axes, out_axes_thunk, donated_invars, global_arg_shapes, *args)
    633                                    in_axes, out_axes_thunk,
    634                                    donated_invars, global_arg_shapes,
--> 635                                    *abstract_args)
    636   return compiled_fun(*args)
    637 

~/Downloads/jax/jax/linear_util.py in memoized_fun(fun, *args)
    249       fun.populate_stores(stores)
    250     else:
--> 251       ans = call(fun, *args)
    252       cache[key] = (ans, fun.stores)
    253 

~/Downloads/jax/jax/interpreters/pxla.py in parallel_callable(fun, backend_name, axis_name, axis_size, global_axis_size, devices, name, in_axes, out_axes_thunk, donated_invars, global_arg_shapes, *avals)
    844              "devices are available (num_replicas={}, num_partitions={})")
    845       raise ValueError(msg.format(num_global_shards, xb.device_count(backend),
--> 846                                   num_global_replicas, num_partitions))
    847 
    848     # On a single host, we use the platform's default device assignment to

ValueError: compiling computation that requires 5 logical devices, but only 2 XLA devices are available (num_replicas=5, num_partitions=1)
```
Note: jax support for amd hardware requires compiling it following hints here: https://github.com/google/jax/issues/2012

**Comment by brianwa84** [comment]
Ah, so it looks like this is a jax usage issue. @sharadmv @mattjj any ideas? Would soft_pmap help here?

**Comment by brianwa84** [comment]
Probably what pymc should be doing is pmap across #devices, then on each device vmap across #chains/#devices. But I think that's what soft_pmap is supposed to do for you.

**Comment by roblem** [comment]
Based on my experience in tensorflow_probability, I usually have more than enough resources on a single gpu even for big models and lots of chains.  Sometimes, I will do simulations or model explorations across gpus using ~joblib~ multiprocessing.pmap, while relying on TF's excellent batching for many chains on a single gpu. 

So what I am saying, is that in my view it should be user definable whether to automatically pmap across all devices (some way of setting visible devices).

**Comment by PaoloRanzi81** [comment]
@roblem 
thanks for confirming that `chain_method=“vectorized”` is forcing the use of CPUs. 
May I ask you two follow-up questions?
1. Could you please try also `pm.sampling_jax.sample_numpyro_nuts()` with `chains=4` + 2 GPUs? This way I can complete the synopsis below. I would expect Numpyro throwing the same error as TFP, though...
2.  Did you notice any increase of RAM when using 2 GPUs concurrently (either by `pm.sampling_jax.sample_tfp_nuts()` or `pm.sampling_jax.sample_numpyro_nuts()` )?
I hope your Bayesian model is complex enough, otherwise it would be very difficult to spot it. If the model is complex, you should see a clear monotonic increase in RAM by naked eye by staring at the output of `htop` command (in a Linux machine).


As follows my 2 benchmarks (1 with 32 vCPUs and 1 with 1 GPU). Of course, I have used the very same data-set and the very same model for all the benchmarks. 

VCPUs
@junpenglao
overall, I can confirm that all 3 PyMC3 functions work very well with vCPUs and 450000 MCMC samples total! Please check table below which summarizes the results. 
I confirm that all 3 functions use the vCPUs “full throttle”, exploiting every single bit of power from the vCPUs. I apologize for the “Minor issue” I have mentioned in a previous comment of mine. Indeed, I was wrong: vCPUs with both `pm.sampling_jax.sample_tfp_nuts()` + `pm.sampling_jax.sample_numpyro_nuts()` are working “full throttle”, as well. Perhaps I was not patient enough to wait for finishing model’s compilation. I judged their performances on vCPUs way too early... 
Anyway, `pm.sampling_jax.sample_numpyro_nuts()` just knocked my socks off… Wow… 2X speed-up with respect to the other two PyMC3 functions!!!

![benchmark_cpus](https://user-images.githubusercontent.com/40964560/107148015-3bb64300-6951-11eb-8c17-970ecb159451.png)


GPUs
@brianwa84 
below the results for `pm.sampling_jax.sample_tfp_nuts()` + 1 GPU. 
I am recalling here the aim I have mentioned in my previous comment: I have tried to get as close as possible to 450000 MCMC samples total with GPUs. `pm.sampling_jax.sample_numpyro_nuts()` was able to get the closest. Indeed, with 1 GPU I was able to get 51000 MCMC samples total with the following combo: `pm.sampling_jax.sample_numpyro_nuts()` + `progress_bar=False` + `chain_method= “vectorized”`. The 51000 MCMC samples total were generated by setting `tune=1000`, `draws=5375` and `chains=8`. 
Unfortunately, `pm.sampling_jax.sample_tfp_nuts()` is not even able to complete a single chain... 

![benchmark_tfp_gpu](https://user-images.githubusercontent.com/40964560/107148030-4cff4f80-6951-11eb-8669-3b539309d38e.png)


TPUs
From a quick reading of the literature and from the above results (successes with vCPUs; failures with GPUs;), it is likely that the problem is to be searched in the breaking down of the MCMC chain. 
Theoretically, TPUs and CPUs should be immune of such an issue since they treat the MCMC chains as a simple embarrassingly parallel problem. Therefore, I would expect 1 TPU working as good as CPUs . However, I have not tested TPUs yet. 

Synopsis of the bench-marking so far:

![synopsis_pymc3_jax](https://user-images.githubusercontent.com/40964560/107148043-58eb1180-6951-11eb-9d79-f8c7998e80b6.png)


Take-home message: 
`pm.sampling_jax.sample_numpyro_nuts()` is very fast with vCPUs!!! GPUs instead are better to be avoided. TPUs should work as good as vCPUs, at least theoretically.

**Comment by roblem** [comment]
>  1. Could you please try also pm.sampling_jax.sample_numpyro_nuts() with chains=4 + 2 GPUs? This way I can complete the synopsis below. I would expect Numpyro throwing the same error as TFP, though...

Numpyro executes 4 chains on a machine with 2 gpus sequentially on only one GPU, which isn't very efficient.   

>  2. Did you notice any increase of RAM when using 2 GPUs concurrently (either by pm.sampling_jax.sample_tfp_nuts() or pm.sampling_jax.sample_numpyro_nuts() )? 

I haven't been able to dig into this, but casual examination of memory usage from the toy models I've been using on this thread may not be large enough to notice and I didn't see anything obvious.

**Comment by PaoloRanzi81** [comment]
@twiecki 

I have tried a TPU V3 by `pm.sampling_jax.sample_tfp_nuts() `+ `pm.sampling_jax.sample_numpyro_nuts()`. Results are not good.

1. Setting TPU in GCP:
The two steps 1.1. + 1.2. below are just a brief summary of a otherwise very complex and convoluted procedure. They can be skipped. I can share the the Python script and Linux commands, in order to save people’s time when undergoing the same procedure. Finally, the TPU setting worked fine (tested with TensorFlow). 

1.1. Setting-up Google Cloud Platform (Compute Engine) by Linux command line:
I did not use any Google Colab for the TPU. I have preferred to manually set up a VM + TPU node in GCP myself. It was a very difficult process since all the tutorials for JAX + TPU are referring to Google Colab. 
The key issue I have found out by trial-and-error, regards the fact that JAX works best with TPU V3, but not with TPU V2. Indeed, TPU V2 gave me lots of problem in installing its drivers as JAX wanted. 
Also, in order to set a VM up and connect it to the TPU node is important to use the TPU’s IP address (adding also “:8470” at the end of the IP address). Using the TPU’s name does not work. 

1.2. Setting-up JAX for working with TPUs by Python script: 
Again, it is not a trivial process in order to get the correct configuration. All the changes must be done at the beginning of the Python script (i.e. before compilation and fitting occur). 
One important issue was to update the TPU’s driver. Then, I had set JAX global variable about the precision to float32 instead of float64 (default in JAX). Not doing that it is guaranteed the TPU throws an error. As I have understood, XLA can not support float64. Furthermore, for some unknown reasons, I could not set JAX’s precision as global variable for bfloat16 (more appropriate for TPUs).  


2. PyMC-JAX results:
Of course I have tried my usual model + usual data-set with PyMC-JAX. 

2.1. `pm.sampling_jax.sample_tfp_nuts()` :
It does not even run. It throws the following error: 

`“RuntimeError: Invalid argument: invalid shape type=4, dims=[-438]”`

2.2. `pm.sampling_jax.sample_numpyro_nuts() `:
Both the summary statistics and the posterior distributions were terrible. Below the first lines of the summary statistics with vCPUs (top) and with TPU (bottom).

![summary_stats_cpu](https://user-images.githubusercontent.com/40964560/107770755-2fa7f800-6d3a-11eb-8de7-4a527436b412.png)

![summary_stats_tpu](https://user-images.githubusercontent.com/40964560/107770843-38003300-6d3a-11eb-9f6f-98b98a94e1d9.png)


You can see the posterior distributions with vCPUs (top) and with TPU (bottom).  

![posteriors_cpu](https://user-images.githubusercontent.com/40964560/107771417-050a6f00-6d3b-11eb-9bfa-1b0c2d174420.png)

![posteriors_tpu](https://user-images.githubusercontent.com/40964560/107771621-4e5abe80-6d3b-11eb-8854-53493bddf9b2.png)


Not sure it is relevant, but with pm.sampling_jax.sample_numpyro_nuts() I have also got repeatedly this warning: 

`“2021-02-11 20:28:30.244877: W external/org_tensorflow/tensorflow/compiler/xla/python/tpu_driver/client/tpu_client.cc:606] TPU Execute is taking a long time. This might be due to a deadlock between multiple TPU cores or a very slow program.”`


3. Synopsis:
I have updated the synopsis according to the more recent -sad- discoveries. 

![synopsis_20210211](https://user-images.githubusercontent.com/40964560/107771104-9200f880-6d3a-11eb-992b-27af8b544340.png)

**Comment by PaoloRanzi81** [comment]
@twiecki 

**1. Good news:** 
1.1. TPU works! In the previous testing with `pm.sampling_jax.sample_numpyro_nuts()` + 1 TPU I had wrongly set a very small “tune”. Now I was able to use `tune = 1633` (no more than that, unfortunately...). The posteriors are equivalent than the ones I have got by vCPUs. Please check tables  below.

![01_tpu_numpyro_tune_1633](https://user-images.githubusercontent.com/40964560/107953316-41d4a100-6f9b-11eb-8b82-72942c57712c.png)

![02_summary_stats_tpu](https://user-images.githubusercontent.com/40964560/107953331-4b5e0900-6f9b-11eb-914e-80e4cfd6cfae.png)

1.2. Luckily, with 1 TPU the monotonic increase of RAM of the Virtual Machine is gone. It seems that only GPUs are affected by such a problem. CPUs and TPUs are not.

1.3. with 1 `TPU pm.sampling_jax.sample_tfp_nuts()` is working now. As for the vCPUs example, with 1 TPU `pm.sampling_jax.sample_tfp_nuts()` is still slower than `pm.sampling_jax.sample_numpyro_nuts()` . 


**2. Bad news:** 
2.1. Very surprisingly, the speed of the TPU is not superior then vCPUs.
I have tried only `tune = 1633`, `draw = 1000`, `chain = 8`. Above `tune = 1633` a “Deadline exceeded” error stopped the TPU (and the TPU must be re-started). 
Please check next point for an explanation of the “Deadline exceeded” error.  
Further, I have estimated that if I had used tune=13062, the TPU would have taken 16212 secs to complete 1 chain. This is in sharp contrast with CPU + `pm.sampling_jax.sample_numpyro_nuts()`: it took 12842 secs. Thus TPU is 26 % slower than CPU. 
Here a quick calculation with the current GCP prices: 
- “e2-highcpu-32” costs $ 0.8/Hr;- 
- “TPU V3” costs $ 8/Hr.
In conclusion, TPU can be cost-effective (therefore beating vCPUs) only if the TPU would take < 22 minutes to run the model. Therefore, vCPUs + `pm.sampling_jax.sample_numpyro_nuts()` looks the best option, as for now. 

2.2. “Deadline exceeded” error: I had this nasty error  which prevented me from running anything which took more than 3600 sec (1 Hr) to fit. That was the reason why I had to stay below tune = 2000 in order to bypass that error message.
The error was present for both `pm.sampling_jax.sample_tfp_nuts()` and `pm.sampling_jax.sample_numpyro_nuts()`. Also, by using `pm.sampling_jax.sample_numpyro_nuts()` the error was still present with either `progress_bar = True` or `progress_bar = False` . 
Even worse: I could not decipher that error. My guess is that Google Cloud Platform switches off the TPU automatically after 1 Hr spent with the TPU not working full-throttle. Does anyone know the meaning for such an error?

![03_deadline_exceeded_error](https://user-images.githubusercontent.com/40964560/107953385-5ca71580-6f9b-11eb-939b-72f7653a4be4.jpg)

2.3. Every 10 seconds a warning is printed saying that the computation is too slow. I have got the same warning even if I had used the the very same tutorial that I have found for “Lorentz attractor” in JAX Github repository. Instead, running TensorFlow this warning does not show up. 

![04_slow_model_warning](https://user-images.githubusercontent.com/40964560/107953475-77798a00-6f9b-11eb-979b-943cca4f6e72.jpg)

2.4. The use that JAX does of the TPU, looks inefficient (specifically the “Utilization of TPU Matrix Units” metric). This is not limited to PyMC-JAX (caption below) because the “Lorentz attractor” tutorial shows the same inefficiency. Instead, TensorFlow does look more efficient.

![05_TPU_utilization](https://user-images.githubusercontent.com/40964560/107953533-8eb87780-6f9b-11eb-92b9-5e036933a254.png)

2.5. Lastly, there are general requirements which should help in capitalizing the TPU’s performance: padding (up to 128 dimension of the matrix); batch size (up to multiple of 64); pre-fetching/caching data directly into the TPU; using bfloat16 data type etc.
Not sure PyMC-JAX, JAX or Numpyro have all of these requirements already implemented. Could be that the reason why TPU is slower than CPU?


**3. Updated synopsis**

![06_synopsis](https://user-images.githubusercontent.com/40964560/107953556-9546ef00-6f9b-11eb-9bfa-0336dead4623.png)

**Comment by twiecki** [comment]
@PaoloRanzi81 Congratulations, you're probably the first person on this planet to have successfully run a PyMC3 model on the TPU :). It's a bummer that it's not faster. My guess is that it indeed needs more careful tuning and debugging, but those probably require more intimate knowledge of how TPUs work, knowledge mainly found inside of Google. Thanks for the update and please keep us posted.

**Comment by martiningram** [comment]
Hi all,

I think I have some things to add here.

A while ago I posted in the [discourse](https://discourse.pymc.io/t/pymc3-jax-negligible-speedup/6447/5) about some changes that seem to have sped things up for me. I've now put together some benchmarks.

The key change I have been looking at is the placement of `jax.jit`. In the current code, `_sample` in `sample_numpyro_nuts` has the `@jax.jit` decorator. Instead, I have been looking at only jit decorating the `log_p_fn_jax`, i.e. the log posterior density function.

I've modified the function with a new argument, `jit_sample_fn`, [here](https://github.com/martiningram/pymc3/blob/fix_numpyro_jit/pymc3/sampling_jax.py#L118). When `jit_sample_fn=True`, the behaviour is the same as currently. When it is False, only the `log_p_fn` has the `jax.jit` decorator. The other change I've made is to allow other chain_methods.

The key benefit is that this fixes `chain_method='vectorized'` on the GPU. When `jit_sample_fn=True`, as reported here, the progress bar runs to 100% but then gets stuck indefinitely. By contrast, with `jit_sample_fn=False`, the progress bar works fine and it runs normally.

I have put together an illustrative example demonstrating this [here](https://colab.research.google.com/drive/1fpzMdEZGC7Zud2rSbyVlvadhKw6UDD5v?usp=sharing). It's a simple logistic regression. Note that the last notebook cell is not expected to finish (this is the problematic current vectorize behaviour). In that basic example, it appears that there is no speed benefit from using `vectorized`, and that when using `parallel`, `jit_sample_fn=True` _does_ accelerate things a little bit.

While this might suggest that `vectorized` isn't worth it, I've found it to give boosts in other models. I've put an example up [here](https://colab.research.google.com/drive/10SzSYDeK-9vXBVi0a9CsWUBxEgO76Jgr?usp=sharing). This is a fairly large hierarchical Bradley-Terry model with correlated random effects. It's essentially a multivariate hierarachical model. Here, `vectorized` takes 10.5 minutes on the GPU, but `parallel` takes 25 minutes. An added benefit is that the progress bar works with `vectorized` but does not with `parallel`. I personally find the progress bar very helpful when fitting models to get an indication of how long it will run.

A final comment is that I had some bugs when using the current versions of `numpyro`, `aesara`, and `jax`. For `jax`, the problem is omnistaging, which is well known I think. Using `0.2.11` is fine. Regarding the bugs with `numpyro` and `aesara`, it's probably best if they go in a separate issue, so I won't go into detail here. The colab notebook has install commands for versions that work.

I hope this is helpful. I can make a pull request with the slight changes to `sample_numpyro_nuts` if people are happy with this.

EDIT: I've gone ahead and made the pull request linked below. Interested to hear people's thoughts!

All the best,
Martin
